Add a Redis extension (#9370)
quarkusio/quarkus: docs/src/main/asciidoc/redis.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////

> = Quarkus - Using Vertx Redis extension

Reviewer: ```suggestion
= Quarkus - Using the Redis Client
```

include::./attributes.adoc[]

> This guide demonstrates how your Quarkus application can use the Redis extension.

Reviewer: ```suggestion
This guide demonstrates how your Quarkus application can connect to a Redis datastore using the Redis extension.
```

[NOTE]
====
This extension is considered `preview`.
API or configuration properties might change as the extension matures.
Feedback is welcome on our https://groups.google.com/d/forum/quarkus-dev[mailing list] or as issues in our https://github.com/quarkusio/quarkus/issues[GitHub issue tracker].

> ====

Reviewer: We have a status include things for that. Have a look at the Hibernate Search Elasticsearch doc.

* Apache Maven 3.5.3+
* A running Redis server, or Docker Compose to start one
* GraalVM installed if you want to run in native mode.

== Architecture

In this guide, we are going to expose a simple Rest API to increment numbers by using the https://redis.io/commands/incrby[`INCRBY`] command.
Along the way, we'll see how to use other Redis commands like `GET`, `SET`, `DEL` and `KEYS`.

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can go right to the completed example.

Clone the Git repository: `git clone {quickstarts-clone-url}`, or download an {quickstarts-archive-url}[archive].

The solution is located in the `redis-quickstart` {quickstarts-tree-url}/redis-quickstart[directory].

== Creating the Maven Project

First, we need a new project. Create a new project with the following command:

[source, subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=redis-quickstart \
    -Dextensions="vertx-redis, resteasy-jsonb, resteasy-mutiny"
cd redis-quickstart
----

> This command generates a Maven project, importing the Redis extension.

Reviewer: See the other guides, we now add a XML snippet with the GA (it's recent so it's normal you didn't add it).

Author: Good idea. 

Done.

----
docker run --ulimit memlock=-1:-1 -it --rm=true --memory-swappiness=0 --name redis_quarkus_test -p 6379:6379 redis:5.0.6
----

== Configuring Redis properties

Once we have the Redis server running, we need to configure the Redis connection properties.
This is done in the `application.properties` configuration file. Edit it to the following content:

[source]
----
quarkus.redis.hosts=localhost:6379 <1>
----

1. Configure Redis hosts to connect to. Here we connect to the Redis server we started in the previous section

== Creating the Increment POJO

We are going to model our increments using the `Increment` POJO.
Create the `src/main/java/org/acme/redis/Increment.java` file, with the following content:

[source, java]
----
package org.acme.redis;

public class Increment {
    public String key; <1>
    public int value; <2>

    public Increment(String key, int value) {
        this.key = key;
        this.value = value;
    }

    public Increment() {
    }
}
----

1. The key that will be used as the Redis key
2. The value held by the Redis key

== Creating the Increment Service

We are going to create an `IncrementService` class which will play the role of a Redis client.

> With this class, we'll be able to do the perform the `SET`, `GET` , `DELET`, `KEYS` and `INCRBY` Redis commands.

Reviewer: ```suggestion
With this class, we'll be able to perform the `SET`, `GET` , `DELET`, `KEYS` and `INCRBY` Redis commands.
```

Create the `src/main/java/org/acme/redis/IncrementService.java` file, with the following content:

[source, java]
----
package org.acme.redis;

import io.quarkus.vertx.redis.SyncRedisAPI;
import io.smallrye.mutiny.Uni;

import io.vertx.mutiny.redis.client.Response;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
class IncrementService {

    @Inject
    SyncRedisAPI syncRedisAPI; <1>

    @Inject

>     io.vertx.mutiny.redis.client.RedisAPI mutinyRedisAPI; <2>

Reviewer: Do we really want the FQCN here?

Author: Surely not. I’ll update this part along with the guide.

    Uni<Void> del(String key) {
        return mutinyRedisAPI.del(Arrays.asList(key))
                .map(response -> null);
    }

    String get(String key) {
        return syncRedisAPI.get(key).toString();
    }

    void set(String key, Integer value) {
        syncRedisAPI.set(Arrays.asList(key, value.toString()));
    }

    void increment(String key, Integer incrementBy) {
        syncRedisAPI.incrby(key, incrementBy.toString());
    }

    Uni<List<String>> keys() {
        return mutinyRedisAPI
                .keys("*")
                .map(response -> {
                    List<String> result = new ArrayList<>();
                    for (Response r : response) {
                        result.add(r.toString());
                    }
                    return result;
                });
    }
}
----

1. Inject the Redis synchronous client

> 2. Inject the mutiny reactive redis client

Reviewer: ```suggestion
2. Inject the Mutiny reactive Redis client
```

                .body("size()", is(2));

        // delete first key
        given()
                .accept(ContentType.JSON)
                .when()
                .delete("/increments/first-key")
                .then()
                .statusCode(204);

        // verify that we have one key left after deletion
        given()
                .accept(ContentType.JSON)
                .when()
                .get("/increments")
                .then()
                .statusCode(200)
                .body("size()", is(1));

        // delete second key
        given()
                .accept(ContentType.JSON)
                .when()
                .delete("/increments/second-key")
                .then()
                .statusCode(204);

        // verify that there is no key left
        given()
                .accept(ContentType.JSON)
                .when()
                .get("/increments")
                .then()
                .statusCode(200)
                .body("size()", is(0));
    }
}
----

== Get it running

If you followed the instructions, you should have the Redis server running.
Then, you just need to run the application using:

[source, shell]
----

> ./mvnw compile quarkus:dev

Reviewer: ```suggestion
./mvnw quarkus:dev
```
compile shouldn't be needed anymore

----

Open another terminal and run the `curl http://localhost:8080/increments` command.

== Interacting with the application
As we have seen above, the API exposes five Rest endpoints.
In this section we are going to see how to initialise an increment, see the list of current increments,
incrementing a value given its key, retrieving the current value of an increment, and finally deleting
a key.

=== Creating a new increment

[source, shell]
----
curl -X POST -H "Content-Type: application/json" -d '{"key":"first","value":10}' http://localhost:8080/increments <1>
----

1. We create the first increment, with the key `first` and an initial value of `10`.

Running the above command should return the result below:

[source, json]
-----
{
  "key": "first",
  "value": 10
}
-----

=== See current increments keys

To see the list of current increments keys, run the following command:

[source, shell]
----
curl http://localhost:8080/increments
----

The above command should return  `["first"]` indicating that we have only one increment thus far.

> === Retrieve an new increment

Reviewer: ```suggestion
=== Retrieve a new increment
```

To retrieve an increment using its key, we will have to run the below command:

[source, shell]
----
curl http://localhost:8080/increments/first <1>
----

1. Running this command, should return the following result:

[source, json]
----
{
  "key": "first",
  "value": 10
}
----

=== Increment a value given its key

To increment a value, run the following command:

[source, shell]
----
curl -X PUT -H "Content-Type: application/json" -d '27' http://localhost:8080/increments/first <1>
----

1. Increment the `first` value by 27.

Now, running the command `curl http://localhost:8080/increments/first` should return the following result:

[source, json]
----
{
  "key": "first",
  "value": 37 <1>
}
----

1. We see that the value of the `first` key is now `37` which is exactly the result of `10 + 27`, quick maths.

=== Deleting a key

Use the command below, to delete an increment given its key.

[source, shell]
----
curl -X DELETE  http://localhost:8080/increments/first <1>
----

1. Delete the `first` increment.

Now, running the command `curl http://localhost:8080/increments` should return an empty list `[]`

> == Packaging and run in JVM mode

Reviewer: ```suggestion
== Packaging and running in JVM mode
```

package org.acme.redis;

public class Increment {
    public String key; <1>
    public int value; <2>

    public Increment(String key, int value) {
        this.key = key;
        this.value = value;
    }

    public Increment() {
    }
}
----

1. The key that will be used as the Redis key
2. The value held by the Redis key

== Creating the Increment Service

We are going to create an `IncrementService` class which will play the role of a Redis client.
With this class, we'll be able to do the perform the `SET`, `GET` , `DELET`, `KEYS` and `INCRBY` Redis commands.

Create the `src/main/java/org/acme/redis/IncrementService.java` file, with the following content:

[source, java]
----
package org.acme.redis;

import io.quarkus.vertx.redis.SyncRedisAPI;
import io.smallrye.mutiny.Uni;

import io.vertx.mutiny.redis.client.Response;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletionStage;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
class IncrementService {

    @Inject
    SyncRedisAPI syncRedisAPI; <1>

    @Inject
    io.vertx.mutiny.redis.client.RedisAPI mutinyRedisAPI; <2>

    @Inject

>     io.vertx.axle.redis.client.RedisAPI axleRedisAPI; <3>

Author: I am seeing that it is deprecated in other extensions e.g mailer. I'll  remove it.

Create the `src/main/java/org/acme/redis/Increment.java` file, with the following content:

[source, java]
----
package org.acme.redis;

public class Increment {
    public String key; <1>
    public int value; <2>

    public Increment(String key, int value) {
        this.key = key;
        this.value = value;
    }

    public Increment() {
    }
}
----

1. The key that will be used as the Redis key
2. The value held by the Redis key

== Creating the Increment Service

We are going to create an `IncrementService` class which will play the role of a Redis client.
With this class, we'll be able to perform the `SET`, `GET` , `DELET`, `KEYS` and `INCRBY` Redis commands.

Create the `src/main/java/org/acme/redis/IncrementService.java` file, with the following content:

[source, java]
----
package org.acme.redis;

import io.quarkus.redis.client.SyncRedisAPI;
import io.smallrye.mutiny.Uni;

import io.vertx.mutiny.redis.client.Response;
import io.vertx.mutiny.redis.client.RedisAPI;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
class IncrementService {

    @Inject

>     SyncRedisAPI syncRedisAPI; <1>

Author: It is sync and blocking since it is a synchronous client. Okay with the renaming to `RedisClient` and I'll also a `ReactiveRedisClient` (which was not there as I expected users to be using the mutiny based one).
